server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,525 @@
+            queueRefs.decRef();
+        }
+    }
+
+    private void drainQueue() {
+        assert isRestResponseFinished.get();
+        assert queueRefs.hasReferences() == false;
+        final var taskCount = queueLength.get() + 1;
+        final var releasables = new ArrayList<Releasable>(taskCount);
+        try {
+            releasables.add(currentEntryReleasable);
+            currentEntryReleasable = null;
+            ChunkedZipEntry entry;
+            while ((entry = entryQueue.poll()) != null) {
+                releasables.add(entry.releasable());
+            }
+            assert entryQueue.isEmpty() : entryQueue.size(); // no concurrent adds
+            assert releasables.size() == taskCount || releasables.size() == taskCount - 1 : taskCount + " vs " + releasables.size();
+        } finally {
+            Releasables.closeExpectNoException(Releasables.wrap(releasables));
+        }
+    }
+
+    /**
+     * A {@link ChunkedRestResponseBodyPart} which will yield all currently-available chunks by consuming entries from {@link #entryQueue}.
+     * There is only ever at most one active instance of this class at any time, in the sense that one such instance becoming inactive
+     * <i>happens-before</i> the creation of the next instance.
+     */
+    private final class AvailableChunksZipResponseBodyPart implements ChunkedRestResponseBodyPart {
+
+        /**
+         * The next {@link ZipEntry} header whose transmission to start.
+         */
+        @Nullable // if no entry is available, or we've already sent the header for the current entry and are now sending its body.
+        private ZipEntry zipEntry;
+
+        /**
+         * The body part which is currently being transmitted, or {@link #NO_MORE_ENTRIES} if we're transmitting the zip file footer.
+         */
+        private ChunkedRestResponseBodyPart bodyPart;
+
+        private boolean isPartComplete;
+        private boolean isLastPart;
+
+        /**
+         * A listener which is created when there are no more available chunks, so transmission is paused, subscribed to in
+         * {@link #getNextPart}, and then completed with the next body part (sequence of zipped chunks, i.e. a new (unique) active
+         * {@link AvailableChunksZipResponseBodyPart}).
+         */
+        private SubscribableListener<ChunkedRestResponseBodyPart> getNextPartListener;
+
+        /**
+         * A cache for an empty list to be used to collect the {@code Releasable} instances to be released when the next chunk has been
+         * fully transmitted. It's a list because a call to {@link #encodeChunk} may yield a chunk that completes several entries, each of
+         * which has its own resources to release. We cache this value across chunks because most chunks won't release anything, so we can
+         * keep the empty list around for later to save on allocations.
+         */
+        private ArrayList<Releasable> nextReleasablesCache = new ArrayList<>();
```

Comment:
> The releaseables are released each time data is sent out, i.e. each time the response is paused or finished at the end.

No, that's not right, the releasables are released when the chunk that completes the entries has been sent, but we do not wait all the way until the end (pause or finish) of the part that contains those entries.

Not sure what else to add to the comments to help here. Although the sentence about caching may not be what you're looking for, the preceding sentence explains why it's a list.

Reply:
Sorry my comment was imprecise. I was meant to say

> The releaseables **are packaged to be released** each time data is sent out, i.e. each time the response is paused or finished at the end.

Essentially it means how we [pass the current list of entry releaseables](https://github.com/elastic/elasticsearch/pull/109820/files#diff-92afd70038ba95143b2feb9d06db9feab773460a9277b20cc63792fc5ef8d44bR393) each time after the do/while loop. 

 > Not sure what else to add to the comments to help here

I think it could help to add a comment right before the do/while loop to say that `writeNextBytes` can work through multiple entries and accumulate entry releaseables which are then released in a single batch once the processed bytes are fully sent out. Feels like a good compliment to the comment here about "a call ... completes serveral entries".